idea/src/org/jetbrains/kotlin/idea/quickfix/AssignOperatorAmbiguityFactory.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.builtins.DefaultBuiltIns
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.diagnostics.Diagnostic
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.descriptorUtil.isSubclassOf
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.KotlinType
+
+object AssignOperatorAmbiguityFactory : KotlinIntentionActionsFactory() {
+    override fun doCreateActions(diagnostic: Diagnostic): List<IntentionAction> {
+        val fixes = mutableListOf<IntentionAction>()
+        val element = diagnostic.psiElement.parent
+        if (element is KtBinaryExpression) {
+            val left = element.left
+            val right = element.right
+            val operationText = when (element.operationToken) {
+                KtTokens.PLUSEQ -> "plus"
+                KtTokens.MINUSEQ -> "minus"
+                else -> null
+            }
+            if (left != null && right != null && operationText != null) {
+                val context = element.analyze(BodyResolveMode.PARTIAL)
+                if (left.getType(context).isMutableCollection()) {
+                    val property = left.mainReference?.resolve() as? KtProperty
+                    val propertyName = property?.name
+                    if (property != null && propertyName != null && property.isLocal) fixes.add(ChangeToValFix(property, propertyName))
+                    fixes.add(ChangeToAssignFunctionCallFix(element, operationText))
+                }
+            }
+        }
+        return fixes
+    }
+}
+
+private fun KotlinType?.isMutableCollection(): Boolean {
```

Comment:
There is `JavaToKotlinClassMap.isMutable(type: KotlinType)`, so you can check the type it self and it's supertypes: `type.constructor.supertypes.any {..}`

Reply:
I don't understand how to use `JavaToKotlinClassMap.isMutable(type: KotlinType)`. `JavaToKotlinClassMap.isMutable(type: KotlinType)` returns `false` when type is ArrayList, HashMap, HashSet.

Reply:
You should check the type it self (for MutableList and etc)  and also its supertypes (for ArrayList and etc)